In [48]:
from rich.jupyter import display

print('hello supervised learning')

hello supervised learning


In [132]:
import polars as pl
import pandas as pd

## Load 

Load from CSV and clean up the data
1. Remove all rows where the year is 2014 because its the first year there are no pbs
2. Clean up gender as it contains "M" or "Male" and "F" or "Female". Just take the first character for simplity
3. Data Types

In [50]:
results_with_pb_df = pl.read_csv("data/withpb_and_age-all-results.csv")


results_with_pb_df = results_with_pb_df.drop("XC_Count_of_event", "XC_Position_Min", "XC_Position_Max", "XC_Min_Age", "XC_Max_Age","exists", "from_file")

# Gender Clean up
results_with_pb_df = results_with_pb_df.with_columns(
    pl.col("Gender").str.slice(0, 1).alias("Gender")
)

results_with_pb_df = results_with_pb_df.with_columns(
    pl.col("is_pb").cast(pl.Boolean),
    pl.col("first_time_run").cast(pl.Boolean)
)

display(results_with_pb_df)




shape: (55_523, 65)
┌──────────┬──────────┬───────────────┬────────────┬───┬─────────────┬─────────┬─────────┬─────────┐
│ Position ┆ Time     ┆ time_in_secon ┆ Bib Number ┆ … ┆ average_age ┆ min_dob ┆ max_dob ┆ avg_dob │
│ ---      ┆ ---      ┆ ds            ┆ ---        ┆   ┆ ---         ┆ ---     ┆ ---     ┆ ---     │
│ i64      ┆ str      ┆ ---           ┆ i64        ┆   ┆ f64         ┆ f64     ┆ f64     ┆ f64     │
│          ┆          ┆ f64           ┆            ┆   ┆             ┆         ┆         ┆         │
╞══════════╪══════════╪═══════════════╪════════════╪═══╪═════════════╪═════════╪═════════╪═════════╡
│ 3        ┆ 00:29:14 ┆ 1754.0        ┆ 880        ┆ … ┆ 42.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 4        ┆ 00:29:34 ┆ 1774.0        ┆ 588        ┆ … ┆ 52.0        ┆ 1964.0  ┆ 1960.0  ┆ 1962.0  │
│ 5        ┆ 00:29:47 ┆ 1787.0        ┆ 902        ┆ … ┆ 37.0        ┆ 1979.0  ┆ 1975.0  ┆ 1977.0  │
│ 6        ┆ 00:30:07 ┆ 1807.0        ┆ 918        ┆ … ┆ 18.5        ┆ 2001.0  ┆ 1990.0  ┆ 1995.5  │
│ 7        ┆ 00:30:09 ┆ 1809.0        ┆ 530        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ …        ┆ …        ┆ …             ┆ …          ┆ … ┆ …           ┆ …       ┆ …       ┆ …       │
│ 627      ┆ 00:47:41 ┆ 2861.0        ┆ 631        ┆ … ┆ 59.0        ┆ 1959.0  ┆ 1955.0  ┆ 1957.0  │
│ 628      ┆ 00:54:49 ┆ 3289.0        ┆ 613        ┆ … ┆ 64.0        ┆ 1954.0  ┆ 1950.0  ┆ 1952.0  │
│ 629      ┆ 00:56:27 ┆ 3387.0        ┆ 429        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 630      ┆ 00:57:56 ┆ 3476.0        ┆ 151        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 631      ┆ 00:58:30 ┆ 3510.0        ┆ 606        ┆ … ┆ 49.0        ┆ 1969.0  ┆ 1965.0  ┆ 1967.0  │
└──────────┴──────────┴───────────────┴────────────┴───┴─────────────┴─────────┴─────────┴─────────┘

## Do we have balanced target classes?

We calculate the PB from the prior runs. 
- The first year is 2014 which has no pbs because there is no prior data
- This needs to be removed as its unbalances the data

We also need to remove any first time only runs as a first is always a pb (technically) and will unbalance the data

In [51]:
results_with_pb_df_all = results_with_pb_df

# Pivot the DataFrame
pivot_df = results_with_pb_df_all.pivot(
    values=["is_pb", "first_time_run"],
    index="year",
    on="is_pb",
    aggregate_function="len"
).sort("year")


display(pivot_df)

# Remove all rows where the year is 2014
results_with_pb_df = results_with_pb_df.filter(
    (pl.col("year") != 2014) & (pl.col("first_time_run") == False)
)

shape: (10, 5)
┌──────┬─────────────┬────────────┬──────────────────────┬─────────────────────┐
│ year ┆ is_pb_false ┆ is_pb_true ┆ first_time_run_false ┆ first_time_run_true │
│ ---  ┆ ---         ┆ ---        ┆ ---                  ┆ ---                 │
│ i64  ┆ u32         ┆ u32        ┆ u32                  ┆ u32                 │
╞══════╪═════════════╪════════════╪══════════════════════╪═════════════════════╡
│ 2014 ┆ 6811        ┆ null       ┆ 6811                 ┆ null                │
│ 2015 ┆ 6289        ┆ 1116       ┆ 6289                 ┆ 1116                │
│ 2016 ┆ 5925        ┆ 1283       ┆ 5925                 ┆ 1283                │
│ 2017 ┆ 5263        ┆ 1042       ┆ 5263                 ┆ 1042                │
│ 2018 ┆ 5269        ┆ 904        ┆ 5269                 ┆ 904                 │
│ 2019 ┆ 5451        ┆ 1172       ┆ 5451                 ┆ 1172                │
│ 2020 ┆ 378         ┆ 93         ┆ 378                  ┆ 93                  │
│ 2022 ┆ 3967        ┆ 637        ┆ 3967                 ┆ 637                 │
│ 2023 ┆ 3758        ┆ 631        ┆ 3758                 ┆ 631                 │
│ 2024 ┆ 4938        ┆ 596        ┆ 4938                 ┆ 596                 │
└──────┴─────────────┴────────────┴──────────────────────┴─────────────────────┘

## Find Categoricals

These are columns with limited subset of values.

For example **education** could be 
- Graduate
- Not Graduate

In our data set
- Club 
- RaceName
- Gender
All of which are unlikely to influence a pb

However, I'll add of these are categorical variables to the model.
- RaceName
- Gender

The two `to_dummies` colum will turn these into columns with 1s and 0s.
- So for `RaceName` it will create a column for each unique value in the column and set it to 1 if the value is present and 0 if not.
- `RaceName_Downland Dash`
- `RaceName_Fittleworth`
- for each race combo





In [52]:
results_with_pb_with_dummies_df = results_with_pb_df.to_dummies(
    ["Gender", "RaceName", 'p_of_10_road_age_category','is_club_member'], drop_first=True
)


In [53]:
display(results_with_pb_with_dummies_df)

shape: (22_443, 102)
┌──────────┬──────────┬───────────────┬────────────┬───┬─────────────┬─────────┬─────────┬─────────┐
│ Position ┆ Time     ┆ time_in_secon ┆ Bib Number ┆ … ┆ average_age ┆ min_dob ┆ max_dob ┆ avg_dob │
│ ---      ┆ ---      ┆ ds            ┆ ---        ┆   ┆ ---         ┆ ---     ┆ ---     ┆ ---     │
│ i64      ┆ str      ┆ ---           ┆ i64        ┆   ┆ f64         ┆ f64     ┆ f64     ┆ f64     │
│          ┆          ┆ f64           ┆            ┆   ┆             ┆         ┆         ┆         │
╞══════════╪══════════╪═══════════════╪════════════╪═══╪═════════════╪═════════╪═════════╪═════════╡
│ 7        ┆ 00:29:11 ┆ 1751.0        ┆ 167        ┆ … ┆ 53.0        ┆ 1964.0  ┆ 1960.0  ┆ 1962.0  │
│ 13       ┆ 00:29:54 ┆ 1794.0        ┆ 203        ┆ … ┆ 28.0        ┆ 1989.0  ┆ 1985.0  ┆ 1987.0  │
│ 19       ┆ 00:30:24 ┆ 1824.0        ┆ 294        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 21       ┆ 00:30:29 ┆ 1829.0        ┆ 390        ┆ … ┆ 43.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 24       ┆ 00:30:45 ┆ 1845.0        ┆ 295        ┆ … ┆ 28.0        ┆ 1989.0  ┆ 1985.0  ┆ 1987.0  │
│ …        ┆ …        ┆ …             ┆ …          ┆ … ┆ …           ┆ …       ┆ …       ┆ …       │
│ 615      ┆ 00:45:24 ┆ 2724.0        ┆ 203        ┆ … ┆ 44.0        ┆ 1974.0  ┆ 1970.0  ┆ 1972.0  │
│ 616      ┆ 00:45:37 ┆ 2737.0        ┆ 112        ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 617      ┆ 00:46:14 ┆ 2774.0        ┆ 63         ┆ … ┆ 34.0        ┆ 1984.0  ┆ 1980.0  ┆ 1982.0  │
│ 623      ┆ 00:49:33 ┆ 2973.0        ┆ 42         ┆ … ┆ null        ┆ null    ┆ null    ┆ null    │
│ 627      ┆ 00:47:41 ┆ 2861.0        ┆ 631        ┆ … ┆ 59.0        ┆ 1959.0  ┆ 1955.0  ┆ 1957.0  │
└──────────┴──────────┴───────────────┴────────────┴───┴─────────────┴─────────┴─────────┴─────────┘

## Split the data

Remove the target column from the features. **is_pb**

Features are the columns that will be used to predict the target variable. 

In this case, the features are all the columns that **Known before the race starts**
- Prior performance 
- Gender
- Club mates (you would see them before the start and on course that could inspire you to run faster)
 


In [83]:
pb_achieved_features_df = results_with_pb_with_dummies_df.drop("is_pb")

# Info about the RACE or runner that is not useful for prediction  
#   runner
pb_achieved_features_df = pb_achieved_features_df.drop("Bib Number", "Name", "Club")
pb_achieved_features_df = pb_achieved_features_df.drop(  'p_of_10_firstname', 'p_of_10_surname', 'p_of_10_track_age_category', 'p_of_10_xc_age_category', 'p_of_10_gender', 'p_of_10_club', 'p_of_10_athlete_id')
#   race
pb_achieved_features_df = pb_achieved_features_df.drop("date", 'year', 'month', 'day', 'distance' )

# achieved time position which would be unknown at the start of the race! 
pb_achieved_features_df = pb_achieved_features_df.drop( "Time", "pace", 'Position', 'time_in_seconds','Gender Position','Points', 'club_position', 'club_position_gender', 'first_time_run' )

# Prior info on years which are versus the current year or just contains the year
pb_achieved_features_df = pb_achieved_features_df.drop( "prior_quickest_run_in_year", "prior_time_vs_current_change", "prior_time_vs_current_time", "prior_quickest_run_time", 'prior_quickest_vs_run_time', 'position_record_best_year', 'position_record_worst_year' )

pb_achieved_features_df = pb_achieved_features_df.drop(  'match_type',  'matches','parameters_used', 'min_age', 'max_age', 'average_age', 'min_dob', 'max_dob', 'avg_dob')

print(pb_achieved_features_df.columns)

display(pb_achieved_features_df)

pb_achieved_target_df = results_with_pb_df.select("is_pb")


['Gender_F', 'is_club_member_false', 'RaceName_Downland Dash', 'RaceName_Fittleworth', 'RaceName_Great Walstead 5', 'RaceName_Gunpowder Trot', 'RaceName_Hangover-5', 'RaceName_Hedgehoppers 5', 'RaceName_Hickstead Gallop', 'RaceName_Highdown Hike', 'RaceName_Hornets Stinger', 'RaceName_Hove Park', 'RaceName_Lewes AC Easter Monday', 'RaceName_Run Your Heart Out 8k', 'RaceName_Seven Stiles', 'RaceName_Steepdown Challenge', 'RaceName_Steyning AC Roundhill Romp', 'RaceName_Tilgate 5', 'RaceName_Trundle View', 'RaceName_Valentine', 'RaceName_Windlesham-Whip', 'RaceName_Worthing Lido', 'prior_run_attempts', 'prior_quickest_run_time_in_seconds', 'prior_quickest_run_position', 'fellow_club_mates_running', 'fellow_club_mates_running_male', 'fellow_club_mates_running_female', 'best_position_year_to_date', 'average_position_year_to_date', 'worst_position_year_to_date', 'best_points_year_to_date', 'average_points_year_to_date', 'worst_points_year_to_date', 'runs_in_wsfrl_year_to_date', 'wsfrl_tenur

shape: (22_443, 61)
┌──────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ Gender_F ┆ is_club_m ┆ RaceName_ ┆ RaceName_ ┆ … ┆ p_of_10_r ┆ p_of_10_r ┆ p_of_10_r ┆ p_of_10_r │
│ ---      ┆ ember_fal ┆ Downland  ┆ Fittlewor ┆   ┆ oad_age_c ┆ oad_age_c ┆ oad_age_c ┆ oad_age_c │
│ u8       ┆ se        ┆ Dash      ┆ th        ┆   ┆ ategory_V ┆ ategory_V ┆ ategory_V ┆ ategory_V │
│          ┆ ---       ┆ ---       ┆ ---       ┆   ┆ 75        ┆ 80        ┆ 85        ┆ 95        │
│          ┆ u8        ┆ u8        ┆ u8        ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│          ┆           ┆           ┆           ┆   ┆ u8        ┆ u8        ┆ u8        ┆ u8        │
╞══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 0        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ …        ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 1        ┆ 0         ┆ 0         ┆ 0         ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
└──────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴───────────┘

In [84]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

SEED = 123

(
    pb_achieved_feature_train_df,
    pb_achieved_feature_test_df,
    pb_achieved_target_train_df,
    pb_achieved_target_test_df,
) = train_test_split(
    pb_achieved_features_df,
    pb_achieved_target_df,
    test_size=0.30,
    random_state=SEED,
)



## Scale the data

It is important to scale the data so that the model is not distorted by really large values.

### My notes are 
- Dimensionality reduction, which 
1. Rotates the data. Some like 45 degrees
2. Keeps the same data but reduce the complexity
- Principal Component Analysis (PCA) is a technique that can be used to reduce the dimensionality of the data.

### Notes from Copilot
The StandardScaler standardizes a feature by subtracting the mean and then scaling to unit variance. Unit variance means dividing all the values by the standard deviation.

In [163]:
scaler = StandardScaler()

pb_achieved_feature_train_scaled_df = scaler.fit_transform(
    pb_achieved_feature_train_df
)

pb_achieved_feature_test_scaled_df = scaler.fit_transform(
    pb_achieved_feature_test_df
)

print(f"Scaled Training data shape: {pb_achieved_feature_train_scaled_df.shape} vs Test {pb_achieved_feature_test_scaled_df.shape}")



Scaled Training data shape: (15710, 61) vs Test (6733, 61)


Do we have balanced target classes?

# Logistic regression example

```
from sklearn.linear_model import LogisticRegression

logistic_regression_classifier = LogisticRegression(penalty=None)

logistic_regression_classifier.fit(
    pb_achieved_feature_train_scaled_df,
    pb_achieved_target_train_df.get_column("is_pb"),
)
```

## Constraints

> ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values


One common strategy is to use the mean or median of the column to fill in the missing values. Here is how you can do it using SimpleImputer from sklearn.impute:  
Import SimpleImputer.
Create an imputer instance with the desired strategy (e.g., mean, median).
Fit the imputer on the training data and transform both the training and test data.

```
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression

# Create an imputer instance with the mean strategy
imputer = SimpleImputer(strategy="mean")

# Fit the imputer on the training data and transform both the training and test data
pb_achieved_feature_train_scaled_df = imputer.fit_transform(pb_achieved_feature_train_scaled_df)
pb_achieved_feature_test_scaled_df = imputer.transform(pb_achieved_feature_test_scaled_df)

# Create and fit the logistic regression classifier
logistic_regression_classifier = LogisticRegression(penalty=None)

logistic_regression_classifier.fit(
    pb_achieved_feature_train_scaled_df,
    pb_achieved_target_train_df.get_column("is_pb"),
)
```





In [118]:
# Convert Polars DataFrame to Pandas DataFrame
pb_achieved_feature_train_df_pandas = pb_achieved_feature_train_df.to_pandas()
pb_achieved_target_train_df_pandas = pb_achieved_target_train_df.to_pandas()
pb_achieved_feature_test_df_pandas = pb_achieved_feature_test_df.to_pandas()
pb_achieved_target_test_df_pandas = pb_achieved_target_test_df.to_pandas()



# Define Column for Defaults

**All columns** need **default values** for **missing values**.

This can have quite an impact on the model. For example if 
- You defaulted missing **points to 10** which is the **highest point** you can score 
- You would **create an artificially high amount** of high points scores when the data is much lower

The second section of the code checks the column exist in the data frame. If they do not exist you will get an error.

> ValueError: A given column is not a column of the dataframe 

In [149]:
# Define the columns
# "Position", "Gender Position", 
time_in_seconds_columns = [ "prior_quickest_run_time_in_seconds"]

years_since_value_columns = ["position_record_years_since_best", "position_record_years_since_worst"]

low_value_good_columns = [
    "best_position_year_to_date", "average_position_year_to_date", "worst_position_year_to_date", "position_record_best_position_all_time", "position_record_worst_position_all_time", "prior_quickest_run_position",  "position_record_average_position_all_time"]

high_value_good_columns = ["fellow_club_mates_running", "fellow_club_mates_running_male", "fellow_club_mates_running_female", "average_points_year_to_date", "best_points_year_to_date", "worst_points_year_to_date", "runs_in_wsfrl_year_to_date", 'wsfrl_tenure_in_runs','wsfrl_tenure_in_years', 'prior_run_attempts']




exclude_races = True
exclude_ages = True

binary_columns = [
    col for col in pb_achieved_feature_train_df_pandas.columns
    if (not exclude_races or not col.startswith('RaceName_')) and 
       (not exclude_ages or not col.startswith('p_of_10_road_age_category_'))
]
binary_columns.append("is_club_member_false")
binary_columns.append("Gender_F")



# Combine the lists
all_feature_columns = set(low_value_good_columns 
                          + high_value_good_columns 
                          + binary_columns 
                          + time_in_seconds_columns 
                          + years_since_value_columns)



# Find columns that are not in the combined list
extra_columns = [col for col in pb_achieved_feature_train_df_pandas.columns if col not in all_feature_columns]
print(f"non range columns in low_value_good_columns: {extra_columns}")

# Check if columns exist in the DataFrame
missing_columns = [col for col in all_feature_columns if col not in pb_achieved_feature_train_df_pandas.columns]
missing_low_value_good_columns = [col for col in low_value_good_columns if col not in pb_achieved_feature_train_df_pandas.columns]
missing_high_value_good_columns = [col for col in high_value_good_columns if col not in pb_achieved_feature_train_df_pandas.columns]


# Print the results
if missing_columns:
    print(f"Missing columns in WHOLE set: {missing_columns}")
else:
    print("All feature columns exist in the DataFrame.")

if missing_low_value_good_columns:
    print(f"Missing columns in low_value_good_columns: {missing_low_value_good_columns}")
else:
    print("All columns in low_value_good_columns exist in the DataFrame.")

if missing_high_value_good_columns:
    print(f"Missing columns in high_value_good_columns: {missing_high_value_good_columns}")
else:
    print("All columns in high_value_good_columns exist in the DataFrame.")
    
    
display(pb_achieved_feature_train_df_pandas.head(4))

# Create a DataFrame with the count of NaN values for each column
nan_counts_df = pd.DataFrame({
    'column': pb_achieved_feature_train_df_pandas.columns,
    'nan_count': pb_achieved_feature_train_df_pandas.isna().sum()
})

# Sort the DataFrame by nan_count in descending order
nan_counts_df = nan_counts_df.sort_values(by='nan_count', ascending=False)

# Set the maximum number of rows to display
pd.set_option('display.max_rows', 30)

nan_counts_df.head(8)

non range columns in low_value_good_columns: ['RaceName_Downland Dash', 'RaceName_Fittleworth', 'RaceName_Great Walstead 5', 'RaceName_Gunpowder Trot', 'RaceName_Hangover-5', 'RaceName_Hedgehoppers 5', 'RaceName_Hickstead Gallop', 'RaceName_Highdown Hike', 'RaceName_Hornets Stinger', 'RaceName_Hove Park', 'RaceName_Lewes AC Easter Monday', 'RaceName_Run Your Heart Out 8k', 'RaceName_Seven Stiles', 'RaceName_Steepdown Challenge', 'RaceName_Steyning AC Roundhill Romp', 'RaceName_Tilgate 5', 'RaceName_Trundle View', 'RaceName_Valentine', 'RaceName_Windlesham-Whip', 'RaceName_Worthing Lido', 'p_of_10_road_age_category_SEN', 'p_of_10_road_age_category_TBC', 'p_of_10_road_age_category_U11', 'p_of_10_road_age_category_U13', 'p_of_10_road_age_category_U15', 'p_of_10_road_age_category_U17', 'p_of_10_road_age_category_U20', 'p_of_10_road_age_category_U23', 'p_of_10_road_age_category_V35', 'p_of_10_road_age_category_V40', 'p_of_10_road_age_category_V45', 'p_of_10_road_age_category_V50', 'p_of_10_

Gender_F  is_club_member_false  RaceName_Downland Dash  \
0         0                     0                       0   
1         0                     1                       0   
2         1                     0                       0   
3         0                     0                       0   

   RaceName_Fittleworth  RaceName_Great Walstead 5  RaceName_Gunpowder Trot  \
0                     0                          0                        0   
1                     0                          0                        0   
2                     0                          0                        0   
3                     0                          0                        0   

   RaceName_Hangover-5  RaceName_Hedgehoppers 5  RaceName_Hickstead Gallop  \
0                    0                        0                          0   
1                    0                        0                          0   
2                    0                        0                          0   
3                    0                        0                          0   

   RaceName_Highdown Hike  ...  p_of_10_road_age_category_V40  \
0                       0  ...                              0   
1                       0  ...                              0   
2                       0  ...                              0   
3                       0  ...                              0   

   p_of_10_road_age_category_V45  p_of_10_road_age_category_V50  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   

   p_of_10_road_age_category_V55  p_of_10_road_age_category_V65  \
0                              1                              0   
1                              0                              0   
2                              0                              0   
3                              1                              0   

   p_of_10_road_age_category_V70  p_of_10_road_age_category_V75  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   

   p_of_10_road_age_category_V80  p_of_10_road_age_category_V85  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   

   p_of_10_road_age_category_V95  
0                              0  
1                              0  
2                              0  
3                              0  

[4 rows x 61 columns]

column  nan_count
average_points_year_to_date      average_points_year_to_date       1626
best_points_year_to_date            best_points_year_to_date       1626
worst_points_year_to_date          worst_points_year_to_date       1626
worst_position_year_to_date      worst_position_year_to_date       1080
best_position_year_to_date        best_position_year_to_date       1080
average_position_year_to_date  average_position_year_to_date       1080
RaceName_Gunpowder Trot              RaceName_Gunpowder Trot          0
RaceName_Hedgehoppers 5              RaceName_Hedgehoppers 5          0

In [150]:
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Create imputers for different columns
# low_value_imputer = SimpleImputer(strategy="constant", fill_value=np.inf)
seconds_value_imputer = SimpleImputer(strategy="constant", fill_value=400000)
low_value_imputer = SimpleImputer(strategy="constant", fill_value=400)
high_value_imputer = SimpleImputer(strategy="constant", fill_value=0) #-np.inf is too big for a float 
binary_value_imputer = SimpleImputer(strategy="constant",fill_value=0)


# Combine the imputers using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ("low_value_good", low_value_imputer, low_value_good_columns),
        ("high_value_good", high_value_imputer, high_value_good_columns),
        ("binary", binary_value_imputer, binary_columns),
        ("seconds", seconds_value_imputer, time_in_seconds_columns),
        ("years_since", SimpleImputer(strategy="constant", fill_value=0), years_since_value_columns),
            
        # ("remaining", SimpleImputer(strategy="constant", fill_value=-1), "passthrough")
    ]
)

PIPELINE_STEP_PREPROCESSOR = "preprocessor"
PIPELINE_STEP_SCALER = "scaler"

# Create a pipeline with the preprocessor, scaler, and the logistic regression classifier
pipeline = Pipeline(steps=[
    (PIPELINE_STEP_PREPROCESSOR, preprocessor),
    (PIPELINE_STEP_SCALER, StandardScaler()),
    ("classifier", LogisticRegression(penalty=None))
])

# Fit the pipeline on the training data
pipeline.fit(pb_achieved_feature_train_df_pandas, pb_achieved_target_train_df_pandas["is_pb"])

# 
# # Transform the test data
# pb_achieved_feature_test_scaled_df = pipeline.transform(pb_achieved_feature_test_df)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('low_value_good',
                                                  SimpleImputer(fill_value=400,
                                                                strategy='constant'),
                                                  ['best_position_year_to_date',
                                                   'average_position_year_to_date',
                                                   'worst_position_year_to_date',
                                                   'position_record_best_position_all_time',
                                                   'position_record_worst_position_all_time',
                                                   'prior_quickest_run_position',
                                                   'position_rec...
                                                   'is_club_member_false',
                                                   'Gender_F']),
                                                 ('seconds',
                                                  SimpleImputer(fill_value=400000,
                                                                strategy='constant'),
                                                  ['prior_quickest_run_time_in_seconds']),
                                                 ('years_since',
                                                  SimpleImputer(fill_value=0,
                                                                strategy='constant'),
                                                  ['position_record_years_since_best',
                                                   'position_record_years_since_worst'])])),
                ('scaler', StandardScaler()),
                ('classifier', LogisticRegression(penalty=None))])

## Run Pipeline Stages explicitly

To run the full pipeline 
`pb_achieved_feature_test_scaled_df = pipeline.transform(pb_achieved_feature_test_df)`

However this does not work 
> AttributeError: This 'Pipeline' has no attribute 'transform'

this is because `LogisticRegression` does not have a `transform` method, but a `fit`. 
  


In [151]:
# Transform the test data using the preprocessor part of the pipeline
pb_achieved_feature_train_preprocessed_df = pipeline.named_steps['preprocessor'].transform(pb_achieved_feature_train_df_pandas)

# Scale the preprocessed test data
pb_achieved_feature_test_scaled_df = pipeline.named_steps['scaler'].transform(pb_achieved_feature_train_preprocessed_df)

In [152]:
from sklearn.linear_model import LogisticRegression

logistic_regression_classifier = LogisticRegression(penalty=None)

logistic_regression_classifier.fit(
    pb_achieved_feature_test_scaled_df,
    pb_achieved_target_train_df_pandas["is_pb"],
)

LogisticRegression(penalty=None)

In [153]:


logistic_regression_classifier.classes_



array([False,  True])

In [154]:
import altair as alt

# Convert the set to a list to maintain order
all_feature_columns_list = list(all_feature_columns)


# Create a DataFrame with the features and their coefficients
feature_coefficients_df = pl.DataFrame(
    {
        "feature": all_feature_columns_list,
        "coefficient": logistic_regression_classifier.coef_[0][:len(all_feature_columns_list)],
    }
)

# Sort the DataFrame by coefficient in descending order
sorted_feature_coefficients_df = feature_coefficients_df.sort("coefficient", descending=True)

# Plot the bar chart
alt.Chart(sorted_feature_coefficients_df.to_pandas()).mark_bar().encode(
    x="coefficient",
    y=alt.Y("feature", sort="-x")
).properties(
    width=500,
    height=300,
).display()

alt.Chart(...)

In [87]:
logistic_regression_classifier.coef_

array([[ 0.02248893, -0.1604215 , -0.35113392,  0.19814715,  0.29265715,
         0.00387152,  0.02259435, -0.01800394,  0.19599449,  0.10406957,
        -0.13826141,  0.30100984, -0.80793519, -0.17211552,  0.19269861,
        -0.00908506, -0.03693306,  0.01017686,  0.02899874,  0.08223687,
         0.01387903,  0.04489769,  0.01410925, -0.07564159, -0.02787106,
         0.02740834,  0.06258948, -0.06138874, -0.049353  , -0.01010949,
         0.21542324, -0.55818381, -0.11816411,  0.04142924]])

In [155]:

# Make predictions on the test data
predictions = pipeline.named_steps['classifier'].predict(pb_achieved_feature_test_scaled_df)

# Print the predictions
print(predictions)

[False False False ... False False False]


In [156]:
print(len(pb_achieved_feature_train_df_pandas))
print(len(pb_achieved_target_train_df_pandas))
print(len(pb_achieved_feature_test_df_pandas))
print(len(pb_achieved_target_test_df_pandas))

15710
15710
6733
6733


In [157]:
print(len(pb_achieved_feature_train_df_pandas.columns))
print(len(pb_achieved_target_train_df_pandas.columns))
print(len(pb_achieved_feature_test_df_pandas.columns))
print(len(pb_achieved_target_test_df_pandas.columns))

61
1
61
1


In [159]:
> What are the probabilities that each of the observations in the test dataset have been assigned to the classes?

Object `classes` not found.


In [158]:
logistic_regression_classifier.predict_proba(pb_achieved_feature_test_scaled_df)

array([[0.97810147, 0.02189853],
       [0.74351853, 0.25648147],
       [0.56647802, 0.43352198],
       ...,
       [0.53827505, 0.46172495],
       [0.50880785, 0.49119215],
       [0.62865966, 0.37134034]])

In [160]:
logistic_regression_classifier.predict(pb_achieved_feature_test_scaled_df)

array([False, False, False, ..., False, False, False])



```
ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values
```

In [176]:
from sklearn.metrics import classification_report


print(
    classification_report(
        pb_achieved_target_test_df.get_column("is_pb"),
        logistic_regression_classifier.predict(pb_achieved_feature_test_scaled_df),
    )
)

ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

In [180]:
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.metrics import classification_report

# Create an imputer instance with the mean strategy
imputer = SimpleImputer(strategy="mean")

# Fit the imputer on the training data and transform both the training and test data
pb_achieved_feature_train_scaled_df = imputer.fit_transform(pb_achieved_feature_train_scaled_df)
pb_achieved_feature_test_scaled_df = imputer.transform(pb_achieved_feature_test_scaled_df)

# Fit the logistic regression classifier again
logistic_regression_classifier.fit(
    pb_achieved_feature_train_scaled_df,
    pb_achieved_target_train_df_pandas["is_pb"],
)

# Generate the classification report
print(
    classification_report(
        pb_achieved_target_test_df_pandas["is_pb"],
        logistic_regression_classifier.predict(pb_achieved_feature_test_scaled_df),
    )
)

              precision    recall  f1-score   support

       False       0.74      0.88      0.80      4468
        True       0.62      0.38      0.47      2265

    accuracy                           0.71      6733
   macro avg       0.68      0.63      0.64      6733
weighted avg       0.70      0.71      0.69      6733



## Interpreting Results from the Classification Report

The classification report provides a summary of performance metrics for a classification model. Below is the detailed breakdown and interpretation:

### Definitions
1. **Precision**: The ratio of true positives to total predicted positives. Indicates how many of the predicted "True" labels are actually "True."
   \[
   \text{Precision} = \frac{\text{True Positives}}{\text{True Positives + False Positives}}
   \]

2. **Recall (Sensitivity)**: The ratio of true positives to total actual positives. Measures the ability of the classifier to identify all "True" instances.
   \[
   \text{Recall} = \frac{\text{True Positives}}{\text{True Positives + False Negatives}}
   \]

3. **F1-Score**: The harmonic mean of precision and recall, balancing their trade-off.
   \[
   \text{F1-Score} = 2 \cdot \frac{\text{Precision} \cdot \text{Recall}}{\text{Precision + Recall}}
   \]

4. **Support**: The number of actual instances in each class.

---

### Interpreting the Results

#### **Class: False (Negative Class)**
- **Precision**: `0.74`  
   - 74% of the predictions for "False" were correct.
- **Recall**: `0.88`  
   - 88% of the actual "False" cases were correctly identified by the model.
- **F1-Score**: `0.80`  
   - A strong balance between precision and recall for this class.
- **Support**: `4468`  
   - There are 4468 instances of "False" in the dataset.

#### **Class: True (Positive Class)**
- **Precision**: `0.62`  
   - 62% of the predictions for "True" were correct.
- **Recall**: `0.38`  
   - Only 38% of the actual "True" cases were correctly identified.
- **F1-Score**: `0.47`  
   - A relatively low score, indicating that the model struggles with this class.
- **Support**: `2265`  
   - There are 2265 instances of "True" in the dataset.

---

### Overall Metrics
- **Accuracy**: `0.71`  
   - The model correctly classified 71% of all instances.
- **Macro Average**:
   - **Precision**: `0.68`  
   - **Recall**: `0.63`  
   - **F1-Score**: `0.64`  
   - These averages give equal weight to both classes, regardless of support.
- **Weighted Average**:
   - **Precision**: `0.70`  
   - **Recall**: `0.71`  
   - **F1-Score**: `0.69`  
   - These averages take into account the support of each class, so the "False" class has more influence due to its larger support.

---

### Insights
1. **Class Imbalance**:  
   - There is a class imbalance, with more "False" instances (`4468`) than "True" (`2265`). This likely affects the model's ability to detect the minority class ("True").

2. **Performance on "True" Class**:
   - The model struggles to correctly identify the "True" class (low recall of `0.38`). This might indicate a need for techniques to handle class imbalance, such as oversampling, undersampling, or class weighting.

3. **Overall Performance**:
   - While the model performs reasonably well on the majority class ("False"), its performance on the minority class ("True") drags down the overall metrics.

---

### Recommendations
1. **Address Class Imbalance**:  
   - Consider using techniques like SMOTE, ADASYN, or class weighting in the logistic regression model to improve recall for the "True" class.

2. **Threshold Tuning**:  
   - Experiment with adjusting the decision threshold of the model to better balance precision and recall.

3. **Model Complexity**:  
   - Consider using a more complex model (e.g., ensemble methods like Random Forest or Gradient Boosting) if the dataset supports it.

4. **Feature Engineering**:  
   - Investigate additional features or transformations to enhance separability between the classes.

---


In [174]:
# Assuming pb_achieved_feature_test_scaled_df is a Pandas DataFrame
# Filter out rows with NaN values
filtered_df = pb_achieved_feature_test_scaled_df.dropna()

# Display the filtered DataFrame
print(filtered_df)

AttributeError: 'numpy.ndarray' object has no attribute 'dropna'